app/models/test_run.rb

```diff
@@ -12,23 +12,33 @@ class TestRun < ApplicationRecord
   STATUSES = {
     complete: 'complete',
     in_progress: 'in_progress',
-    cancelled: 'cancelled'
+    cancelled: 'cancelled',
+    complete_with_errors: 'complete_with_errors'
   }.freeze
 
   def status
-    return STATUSES[:complete] unless test_script_results.empty?
+    unless test_script_results.empty?
```

Comment:
The structure here is pretty funky. It would be good to change this to a flatter conditional:

```
if test_script_results.(with an error).exists?
  STATUSES[:complete_with_errors]
elsif test_script_results.exists?
  STATUSES[:complete]
elsif time_to_service&.negative?
  STATUSES[:cancelled]
else
  STATUSES[:in_progress]
end
```

Reply:
I actually thing the nested if statements make sense in this context. The `complete_with_errors` status is only ever assigned to test runs we would otherwise give a `complete` status to, so it makes sense to treat one as a subset of the other in the logic.  

Also this mirrors the way I have structured the logic in the `TestRun.status` method.  If you want, I can do something like


      if test_script_results_id
        if completion_status == 'error' || status_hash[id] == STATUSES[:complete_with_errors]
          status_hash[id] = STATUSES[:complete_with_errors]
          next
        end
        status_hash[id] = STATUSES[:complete]
      elsif time_to_service&.negative?
        status_hash[id] = STATUSES[:cancelled]
      else
        status_hash[id] = STATUSES[:in_progress]
      end

To make it a little more clear that the `complete` status is the default in this case and the `complete_with_errors` is a special case.